### ISSUE TITLE
Cannot upgrade/downgrade legacy addons

### ISSUE THREAD

**Issue by zunda** [heroku/heroku-addon-attachments#28]
We have a difficulty in upgrading an addon with heroku-addon-attachments
 plugin installed:

- I suppose we should align the usage with toolbelt without this plugin. 
- I don't know which side of the API has to be modified.

```
$ heroku addons:add proximo
$ heroku addons:upgrade proximo:starter
Upgrading proximo:starter to ... failed
 !    Couldn't find that add-on.
```

On the client side, the problem seems to be that the usage is different from when heroku-addon-attachments is not installed. I made a modification as in  cannot-find-add-on branch @cc165802bae2e53bfb715df087e0d11b497628d6 but it seems that access to the API also has to be modified:

```
$ heroku addons:upgrade proximo:starter
Upgrading proximo to starter... failed
 !    No such add-on plan.
```

**Comment by bjeanes** [comment]
Hey @zunda,

The `heroku-addon-attachments` plugin works within a representation of add-ons that is N:N with apps, where they used to be 1:1. As such, the previous invocation style of `heroku addons:upgrade NEW_PLAN` is ambiguous in the case where you have multiple installations of an add-on.

As per the help text for that command, the invocation now is `heroku addons:upgrade ADDON PLAN`, where `ADDON` is the globally unique name for that add-on instance (which can be seen in `heroku addons` output).

In theory, we could proceed as the CLI previously did in unambiguous cases but we are erring on the side of consistency for now. For sure, a middle ground might exist, such as:

``` 
$ heroku addons:upgrade proximo:starter
Missing add-on identifier. Did you mean to upgrade `my-add-on-123`?
Run `heroku addons:upgrade my-add-on-123 proximo:starter` instead.
```

What do you think?

**Comment by zunda** [comment]
Hi, 

I am actually a Support Engineer at Heroku. Thanks for the explanation about handling the attachable addons. I totally forgot about this.

The problem I see supporting customers is that sometimes they install this plugin and follow legacy documentations. In this regard, it is the best if the plugin does not change behaviour of the command for legacy addons. I am afraid that I still need time to understand the whole picture of attachable addons so I would like to take some more time to study about the coming add-on before I can see what is the best way to go.

**Comment by bjeanes** [comment]
> The problem I see supporting customers is that sometimes they install this plugin and follow legacy documentations. 

We have to change the invocation. We are changing it in other places anyway. The docs will need to be updated once this plugin is merged into the main toolbelt.

**Comment by ddollar** [comment]
This keeps biting my customers. It looks to them like down/upgrade is broken every time they try to use it.

**Comment by bjeanes** [comment]
@ddollar what do you think of my idea above for a change in the output text in these situations?

**Comment by ddollar** [comment]
It's ok but it puts the onus back on the user for no real reason. If they only specify one argument just assume they're trying to work with a legacy addon. Im assuming you can even do some magic with the API in the backend to know if they're specifying a "legacy" addon.

**Comment by bjeanes** [comment]
@ddollar the thing is that there is no more "legacy" add-on. This time around shareable add-ons we've consolidated the entire data model. I think having varying and inconsistent invocations is worse, personally. We certainly *could* infer that the customer meant some specific installation if that add-on hasn't been configured to allow multiple installations per app (which is externally much like a "legacy" add-on) but I suspect this will be just as surprising and opaque when we can't infer it on other add-ons.

What do you think?

**Comment by ddollar** [comment]
If someone installs this plugin suddenly many Devcenter docs are invalid for them. I think this is a problem. The rest seems theoretical. Make the CLI do whatever crazy nonsense is necssary in order to prevent the user from having to worry about or make decisions about a platform implementation detail they couldn't care less about.

**Comment by bjeanes** [comment]
@ddollar Alright I'll think about this a bit more. I am very hesitant to introduce inconsistency but at least for a grace period it might be reasonable to make the old invocation work with some explanatory text. I'll update here with any conclusions or changes. Thanks for thinking it through with us.

**Comment by bjeanes** [comment]
@ddollar I just pushed up #34 with some WIP for handling this client-side when we can unambiguously resolve the add-on from just the app and plan. More testing needs to be done when it *is* ambiguous and the messaging certainly needs work. 

What do you think?

``` sh-session
$ heroku addons:upgrade newrelic:wayne -a bjeanes
WARNING: No add-on name specified (see `heroku help addons:upgrade`)
Finding add-on from service newrelic on app bjeanes... done
Found unwinding-kindly-4330 (newrelic:stark) on bjeanes.
Upgrading unwinding-kindly-4330 to newrelic:wayne... done
```

**Comment by bjeanes** [comment]
I was hoping to some feedback on #34 from the people in this thread but I've gone ahead and merged it now anyway with some further local testing.

**Comment by ddollar** [comment]
I would personally avoid all of the extra output there and just make it look exactly like the old flow. It feels like the migration from "legacy" (odd word since as a provider I can't even update to the "current" model if I wanted to right now) to "current" should be an implementation detail that could be completely masked by Heroku.

In short, why use 4 lines when you can use 1?

That being said, thanks for fixing it. :) This is certainly an improvement over not working at all.

**Comment by bjeanes** [comment]
I don't believe I've used the world legacy other than quoting the initial title in this context. Internally we are referring to these add-ons as "classic" add-ons. In any case, there isn't a distinction in the model, just in the validations for whether multiple attachments can occur. Classic add-ons are just attachable add-ons that can only be attached once to its billing app. Implementation details are not important to the customer, and I understand that, but I believe internal consistency. 

> In short, why use 4 lines when you can use 1?

I want this reference style to be temporary. It's done client-side so it's slower and making APIv3 look up by plan identifiers in the :id param is off the table. I think reducing it to the warning + a single line is worth playing with though. 

I route the lookup function in a way that we can start applying it to the other commands too for a more graceful upgrade experience.
